File: LeetCode 224- Basic caculator/digits_+-()/basic_calculator_I_stack.py
class Solution(object):
     def calculate(self, s):
        """
        :type s: str
        :rtype: int
        """
        result = 0
        operand = 0
        operator = 1 # "+"
        stack = []
        
        for char in s:
            if '0' <= char <= '9':
                operand = operand * 10 + int(char)
            elif char in "-+":
                result += operand * operator
                # reset operator
                operand = 0
                # update operator
                operator = 1 if char == '+' else -1
            elif char == "(":
                stack.append(result)
                stack.append(operator)
                # reset result and operator
                result = 0
                operator = 1
            elif char == ")":
                result += operand * operator
                result *= stack.pop()
                result += stack.pop()
                # reset operand
                operand = 0
            
        result += operand * operator
                
        return result

if __name__ == "__main__":
    solution = Solution()
    s = "10+(2-3+5)-6"
    print("result", solution.calculate(s))

"""
output:
('result', 8)
"""

File: LeetCode 224- Basic caculator/digits_+-()/basic_calculator_I_stack.go
package main

import "fmt"

func calculate(s string) int {
	result := 0
	operand := 0
	operator := 1 // +
	stack := []int{}

	for _, char := range s {
		if '0' <= char && char <= '9' {
			operand = operand*10 + int(char-'0')
		} else if char == '-' || char == '+' {
			result += operand * operator
			// reset operand
			operand = 0
			// update operator
			operator = 1
			if char == '-' {
				operator = -1
			}
		} else if char == '(' {
			stack = append(stack, result)
			stack = append(stack, operator)
			// reset result and operator
			result = 0
			operator = 1
		} else if char == ')' {
			result += operand * operator
			result *= stack[len(stack)-1]
			result += stack[len(stack)-2]
			stack = stack[:len(stack)-2]
			// reset operand
			operand = 0
		}
	}

	result += operand * operator
	return result
}

func main() {
	s := "10+(2-3+5)-6"
	fmt.Println("result:", calculate(s))
}

/*
output:
result: 8
*/
